translator/metrics/otlp_to_oc.go

```diff
@@ -0,0 +1,377 @@
+		Resource: resource,
+	}
+
+	if len(resourceMetrics.Metrics) == 0 {
+		return md
+	}
+
+	metrics := make([]*ocmetrics.Metric, 0, len(resourceMetrics.Metrics))
+	for _, metric := range resourceMetrics.Metrics {
+		metrics = append(metrics, metricToOC(metric))
+	}
+
+	md.Metrics = metrics
+	return md
+}
+
+func metricToOC(metric *otlpmetrics.Metric) *ocmetrics.Metric {
+	labelKeys := allLabelKeys(metric)
+	return &ocmetrics.Metric{
+		MetricDescriptor: descriptorToOC(metric.MetricDescriptor, labelKeys),
+		Timeseries:       dataPointsToTimeseries(metric, labelKeys),
+		Resource:         nil,
+	}
+}
+
+func descriptorToOC(descriptor *otlpmetrics.MetricDescriptor, labelKeys []*ocmetrics.LabelKey) *ocmetrics.MetricDescriptor {
+	if descriptor == nil {
+		return nil
+	}
+
+	return &ocmetrics.MetricDescriptor{
+		Name:        descriptor.Name,
+		Description: descriptor.Description,
+		Unit:        descriptor.Unit,
+		Type:        descriptorTypeToOC(descriptor.Type),
+		LabelKeys:   labelKeys,
+	}
+}
+
+func allLabelKeys(metric *otlpmetrics.Metric) []*ocmetrics.LabelKey {
+	// NOTE: OpenTelemetry and OpenCensus have different representations of labels:
+	// - OC has a single "global" ordered list of label keys per metric in the MetricDescriptor;
+	// then, every data point has an ordered list of label values matching the key index.
+	// - In OTLP, every label is represented independently via StringKeyValue,
+	// i.e. theoretically points in the same metric may have different set of labels.
+	//
+	// So what we do in this translator:
+	// - Scan all points and their labels to generate a unique set of all label keys
+	// used across the metric, sort them and set in MetricDescriptor.
+	// - For each point we generate an ordered list of label values,
+	// matching the order of label keys returned here (see `labelsToOC` function).
+
+	// First, collect a set of unique keys
+	uniqueKeys := make(map[string]struct{}, 0)
+	for _, point := range metric.Int64Datapoints {
+		addLabelKeys(uniqueKeys, point.Labels)
+	}
+	for _, point := range metric.DoubleDatapoints {
+		addLabelKeys(uniqueKeys, point.Labels)
+	}
+	for _, point := range metric.HistogramDatapoints {
+		addLabelKeys(uniqueKeys, point.Labels)
+	}
+	for _, point := range metric.SummaryDatapoints {
+		addLabelKeys(uniqueKeys, point.Labels)
+	}
+
+	// Sort keys (this is actually optional, just for stable order)
+	rawKeys := make([]string, 0, len(uniqueKeys))
+	for key := range uniqueKeys {
+		rawKeys = append(rawKeys, key)
+	}
+	sort.Strings(rawKeys)
+
+	// Construct a list of label keys
+	// Note: Label values will have to match keys by index
+	labelKeys := make([]*ocmetrics.LabelKey, 0, len(rawKeys))
+	for _, key := range rawKeys {
+		labelKeys = append(labelKeys, &ocmetrics.LabelKey{
+			Key: key,
+		})
+	}
+
+	return labelKeys
+}
+
+func addLabelKeys(uniqueKeys map[string]struct{}, labels []*otlpcommon.StringKeyValue) {
+	for _, label := range labels {
+		if _, ok := uniqueKeys[label.Key]; !ok {
```

Comment:
Do we need to check for existence before setting? This appears to be unnecessary and will likely be slower.